cw-orch/src/interchain/follow_ibc_execution.rs

```diff
@@ -0,0 +1,211 @@
+///     - The chain id of the destination chain is not available directly in the logs.
+///         However, it is possible to query the node for the chain id of the counterparty chain linked by a connection
+///
+/// 2. Receive packet. For each packet received, we use the identification of the packet that was sent in the original tx to find the tx in which the packet was received
+///     We make sure that only one transaction tracks receiving this packet.
+///         If not, we sent out an error (this error actually comes from the code not identifying an IBC packet properly)
+///         If such an error happens, it means this function is not implemented properly
+///         We verify this transaction is not errored (it should never error)
+///     
+/// 3. Acknowledgment. The last part of the packet lifetime is the acknowledgement the distant chain sents back.
+///     a. Identify acknowledgement
+///         In the same transaction as the one in which the packet is received, an packet acknowledgement should be sent back to the origin chain
+///         We get this acknowledgment and deserialize it according to https://github.com/cosmos/cosmos-sdk/blob/v0.42.4/proto/ibc/core/channel/v1/channel.proto#L134-L147
+///         If the acknowledgement doesn't follow the standard, we don't mind and continue
+///     b. Identify the acknowledgement receive packet on the origin chain
+///         Finally, we get the transaction hash of the transaction in which the acknowledgement is received on the origin chain.
+///         This is also logged for debugging purposes
+///
+/// 4. Finally, some additionnal packets may have been sent out during the whole process. We need to check the generated transactions for additional packets.
+///    This 4th step make the whole process recursive. Those two transactions potentially sends out additional packets !
+///         - The receive Packet transaction identified in 2. on the distant chain
+///         - The receive Acknowledgement transaction identified in 3.b. on the origin chain
+
+#[async_recursion::async_recursion]
+pub async fn follow_trail(chain1: String, channel1: Channel, tx_hash: String) -> Result<()> {
+    // 1. Getting IBC related events for the current tx
+    let tx = Node::new(channel1.clone()).find_tx(tx_hash.clone()).await?;
+
+    let send_packet_events = tx.get_events("send_packet");
+    if send_packet_events.is_empty() {
+        return Ok(());
+    }
+
+    log::info!(
+        target: &chain1,
+        "Investigating sent packet events on tx {}",
+        tx_hash
+    );
+    let connections = get_events(&send_packet_events, "packet_connection");
+    let src_ports = get_events(&send_packet_events, "packet_src_port");
+    let src_channels = get_events(&send_packet_events, "packet_src_channel");
+    let sequences = get_events(&send_packet_events, "packet_sequence");
+    let packet_datas = get_events(&send_packet_events, "packet_data");
+    let chain_ids = try_join_all(
```

Comment:
I'm pretty sure the connections, src_port, channel sequence can all only be one value. It's a vec because of the way that attributes are stored. 

Could you refactor the code to use single values instead of the vectors?

Reply:
No, if a transaction sends multiple packets, there will be multiple attributes with the same name :)